### PR TITLE
Automatically reload tenant when configuration is changed.

### PR DESCRIPTION
**Pull request by MatthijsKrempel**
Fixes https://github.com/OrchardCMS/OrchardCore/issues/4461

Reload tenant when settings are updated.

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.ReverseProxy/Drivers/ReverseProxySettingsDisplayDriver.cs [hunk 1]
```diff
@@ -15,63 +16,56 @@ namespace OrchardCore.ReverseProxy.Drivers
...
-        private readonly INotifier _notifier;
-        private readonly IMemoryCache _memoryCache;
+        private readonly IShellHost _orchardHost;
+        private readonly ShellSettings _currentShellSettings;
 
         public ReverseProxySettingsDisplayDriver(INotifier notifier,
-            IMemoryCache memoryCache,
+            IShellHost orchardHost,
+            ShellSettings currentShellSettings,
             IHtmlLocalizer<ReverseProxySettingsDisplayDriver> stringLocalizer)
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.ReverseProxy/Drivers/ReverseProxySettingsDisplayDriver.cs]: Do you need to remove `INotifier` from here as well, now it's unused?

Going to refer you to https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Modules/OrchardCore.Https/Drivers/HttpsSettingsDisplayDriver.cs

for the naming conventions that are used everywhere else, for restarting tenants
ie. `shellHost`, and `shellSettings`

File: src/OrchardCore.Modules/OrchardCore.ReverseProxy/Drivers/ReverseProxySettingsDisplayDriver.cs [hunk 2]
```diff
@@ -15,63 +16,56 @@ namespace OrchardCore.ReverseProxy.Drivers
...
-            _memoryCache = memoryCache;
+            _orchardHost = orchardHost;
+            _currentShellSettings = currentShellSettings;
             T = stringLocalizer;
         }
         IHtmlLocalizer T { get; }
 
-        public override IDisplayResult Edit(ReverseProxySettings settings, BuildEditorContext context)
+        public override IDisplayResult Edit(ReverseProxySettings section, BuildEditorContext context)
         {
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.ReverseProxy/Drivers/ReverseProxySettingsDisplayDriver.cs]: `settings` is used everywhere else

File: src/OrchardCore.Modules/OrchardCore.ReverseProxy/Drivers/ReverseProxySettingsDisplayDriver.cs [hunk 3]
```diff
@@ -15,63 +16,56 @@ namespace OrchardCore.ReverseProxy.Drivers
...
 
             }
 
-            if (_memoryCache.Get(RestartPendingCacheKey) == null)
-            {
-                var entry = _memoryCache.CreateEntry(RestartPendingCacheKey);
-                _memoryCache.Set(entry.Key, entry, new MemoryCacheEntryOptions { Priority = CacheItemPriority.NeverRemove });
-            }
+            // Reload the tenant to apply the settings
+            await _orchardHost.ReloadShellContextAsync(_currentShellSettings);
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.ReverseProxy/Drivers/ReverseProxySettingsDisplayDriver.cs]: Need to wrap this to only reset site if the settings are good
```
                // If the settings are valid, reload the current tenant.
                if (context.Updater.ModelState.IsValid)
                {
                    await _shellHost.ReloadShellContextAsync(_shellSettings);
                }
```

File: src/OrchardCore.Modules/OrchardCore.ReverseProxy/Views/ReverseProxySettings.Edit.cshtml [hunk 4]
```diff
@@ -33,6 +33,14 @@
...
-<span class="hint">
-    Documentation available at; https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-2.1
-</span>
\ No newline at end of file
+<div class="row">
+    <span class="hint">
+        @T["The current tennant will be reloaded when settings are updated."]
+    </span>
+</div>
+
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.ReverseProxy/Views/ReverseProxySettings.Edit.cshtml]: The warning is normally at the top of the page.

https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Modules/OrchardCore.Https/Views/HttpsSettings.Edit.cshtml
for example.

Plus then the text is the same as elsewhere when people are doing translations.
```
<p class="alert alert-warning">@T["The current tenant will be reloaded when the settings are saved."]</p>
```

File: src/OrchardCore.Modules/OrchardCore.ReverseProxy/Drivers/ReverseProxySettingsDisplayDriver.cs (hunk 2 above)
Comment (MatthijsKrempel) [reply src/OrchardCore.Modules/OrchardCore.ReverseProxy/Drivers/ReverseProxySettingsDisplayDriver.cs]: Then we should change the interface, my implementation adheres to the interface ;)

File: src/OrchardCore.Modules/OrchardCore.ReverseProxy/Drivers/ReverseProxySettingsDisplayDriver.cs (hunk 3 above)
Comment (MatthijsKrempel) [reply src/OrchardCore.Modules/OrchardCore.ReverseProxy/Drivers/ReverseProxySettingsDisplayDriver.cs]: Will do

File: src/OrchardCore.Modules/OrchardCore.ReverseProxy/Drivers/ReverseProxySettingsDisplayDriver.cs (hunk 2 above)
Comment (deanmarcussen) [reply src/OrchardCore.Modules/OrchardCore.ReverseProxy/Drivers/ReverseProxySettingsDisplayDriver.cs]: As you wish. Probably something we should clean up, but another pr later.
About 80% use `settings` and the others `section`

Other than that all LGTM

**Comment by deanmarcussen** [comment]
@agriffard if you are resolving conflicts and merging this LGTM.

**Pull request merged**